### ISSUE TITLE
Installation of nloptr fails without internet

### ISSUE THREAD

**Issue by karthik2527** [astamm/nloptr#74]
I am trying to install nloptr_1.0.4.tar.gz (had to be old version due to user requirement) during compilation of the package it tries to download from `http://ab-initio.mit.edu/nlopt/nlopt-2.4.2.tar.gz` since our server doesn't have internet connection it fails after trying for sometime. Is there a way we could install this package without having internet connection please ? 

OS `Red Hat Enterprise Linux release 8.3 (Ootpa)`
R version `R version 3.4.3 (2017-11-30) -- "Kite-Eating Tree"`

**Comment by eddelbuettel** [comment]
The installation is already capable of detecting an already existing installation of NLopt in which case _no download_ is needed or made:

```sh
edd@rob:~$ install.r nloptr       # simple wrapper around install.packages()
trying URL 'https://cloud.r-project.org/src/contrib/nloptr_1.2.2.2.tar.gz'
Content type 'application/x-gzip' length 2450931 bytes (2.3 MB)
==================================================
downloaded 2.3 MB

* installing *source* package ‘nloptr’ ...
** package ‘nloptr’ successfully unpacked and MD5 sums checked
** using staged installation
checking whether the C++ compiler works... yes
checking for C++ compiler default output file name... a.out
checking for suffix of executables... 
checking whether we are cross compiling... no
checking for suffix of object files... o
checking whether we are using the GNU C++ compiler... yes
checking whether ccache g++ accepts -g... yes
checking how to run the C++ preprocessor... ccache g++ -E
checking whether we are using the GNU C++ compiler... (cached) yes
checking whether ccache g++ accepts -g... (cached) yes
checking for pkg-config... yes                                      # Step One
configure: Now testing for NLopt header file.
checking for grep that handles long lines and -e... /bin/grep
checking for egrep... /bin/grep -E
checking for ANSI C header files... yes
checking for sys/types.h... yes
checking for sys/stat.h... yes
checking for stdlib.h... yes
checking for string.h... yes
checking for memory.h... yes
checking for strings.h... yes
checking for inttypes.h... yes
checking for stdint.h... yes
checking for unistd.h... yes
checking nlopt.h usability... yes                                    # Step Two
checking nlopt.h presence... yes                                     # Step Three
checking for nlopt.h... yes                                          # Step Four
configure: Now testing for NLopt version 2.4.0 or greater.
configure: Suitable NLopt library found.
configure: creating ./config.status
config.status: creating src/Makevars
config.status: creating R/PkgFlags.R
** libs
ccache g++ -I"/usr/share/R/include" -DNDEBUG      -fpic  -g -O3 -Wall -pipe -pedantic  -c dummy.cpp -o dummy.o
ccache gcc -I"/usr/share/R/include" -DNDEBUG      -fpic  -g -O3 -Wall -pipe -pedantic  -std=gnu99 -c init_nloptr.c -o init_nloptr.o
ccache gcc -I"/usr/share/R/include" -DNDEBUG      -fpic  -g -O3 -Wall -pipe -pedantic  -std=gnu99 -c nloptr.c -o nloptr.o
ccache g++ -Wl,-S -shared -L/usr/lib/R/lib -Wl,-Bsymbolic-functions -Wl,-z,relro -o nloptr.so dummy.o init_nloptr.o nloptr.o -lnlopt -L/usr/lib/R/lib -lR
installing to /usr/local/lib/R/site-library/00LOCK-nloptr/00new/nloptr/libs
** R
** inst
** byte-compile and prepare package for lazy loading
** help
*** installing help indices
** building package indices
** installing vignettes
** testing if installed package can be loaded from temporary location
** checking absolute paths in shared objects and dynamic libraries
** testing if installed package can be loaded from final location
** testing if installed package keeps a record of temporary installation path
* DONE (nloptr)

The downloaded source packages are in
	‘/tmp/downloaded_packages’
edd@rob:~$ 
```

I highlighted above where some of the tests happen, as you can see, no download is attempted, made or needed -- here we just link with `-lnlopt` (as my system has it in a system location).

So I suggest we close this.

**Comment by karthik2527** [comment]
thanks could you please advice if NLopt is available for Centos 8 please ?

**Comment by astamm** [comment]
We just released `nloptr v2.0.0` on CRAN. Binaries for Windows and macOS are on their way. I understand you are using Linux CentOS. In this case, `nloptr` will look for an existing system build of NLopt on your machine of recent enough version (`>= 2.7.0`). If it finds one, it will use it. Otherwise, it will build NLopt 2.7.1 from included sources using `cmake`. To do so, it will attempt to find `cmake` binary on your `PATH`. If it fails, it will display a list of ways to get `cmake` installed on your machine.
In general, on Linux, macOS and Windows-devel, there is no download of NLopt whatsoever now. The only case where we still need an internet connexion is the current Windows release for which we download NLopt 2.7.1 from `rwinlib`.
